blocks/address/src/app.tsx

```diff
@@ -0,0 +1,677 @@
+
+  const decrementZoomLevel = useCallback(() => {
+    if (zoomLevel >= MIN_ZOOM_LEVEL + ZOOM_LEVEL_STEP_SIZE) {
+      updateZoomLevel(zoomLevel - ZOOM_LEVEL_STEP_SIZE);
+    }
+  }, [zoomLevel, properties]);
+
+  const uploadMap = async (mapFile: File) => {
+    if (readonly || !mapFile) {
+      return;
+    }
+
+    graphService
+      ?.uploadFile({
+        data: { file: mapFile },
+      })
+      .then(async (uploadFileResponse) => {
+        const imageUrl =
+          uploadFileResponse.data?.properties[
+            "https://blockprotocol.org/@blockprotocol/types/property-type/url/"
+          ];
+
+        if (imageUrl) {
+          const fileProperties = {
+            [imageUrlKey]: imageUrl,
+          };
+
+          const createFileEntityResponse = await graphService?.createEntity({
+            data: {
+              entityTypeId: fileTypeId,
+              properties: fileProperties,
+            },
+          });
+
+          const fileEntityId =
+            createFileEntityResponse?.data?.metadata.recordId.entityId;
+
+          if (!imageLinkEntity && fileEntityId && addressId) {
+            await graphService?.createEntity({
+              data: {
+                entityTypeId: imageLinkTypeId,
+                properties: {
+                  [zoomLevelKey]: zoomLevel,
+                  [addressIdKey]: addressId,
+                },
+                linkData: {
+                  leftEntityId: entityId,
+                  rightEntityId: fileEntityId,
+                },
+              },
+            });
+          }
+        }
+      });
+  };
+
+  const updateAddress = async (address?: Address) => {
+    if (readonly || !address) {
+      return;
+    }
+
+    const {
+      addressLocality,
+      addressRegion,
+      postalCode,
+      streetAddress,
+      addressCountry,
+      fullAddress,
+    } = address;
+
+    const addressProperties = {
+      [localityKey]: addressLocality,
+      [regionKey]: addressRegion,
+      [postalCodeKey]: postalCode,
+      [streetKey]: streetAddress,
+      [countryKey]: addressCountry,
+      [fullAddressKey]: fullAddress,
+    };
+
+    const createAddressEntityResponse = await (!addressEntity
+      ? graphService?.createEntity({
+          data: {
+            entityTypeId: addressTypeId,
+            properties: addressProperties,
+          },
+        })
+      : graphService?.updateEntity({
+          data: {
+            entityId: addressEntity.metadata.recordId.entityId,
+            entityTypeId: addressEntity.metadata.entityTypeId,
+            properties: addressProperties,
+          },
+        }));
+
+    const addressEntityId =
+      createAddressEntityResponse?.data?.metadata.recordId.entityId;
+
+    if (addressEntityId) {
+      if (!addressLinkEntity) {
+        await graphService?.createEntity({
+          data: {
+            entityTypeId:
+              "https://alpha.hash.ai/@luisbett/types/entity-type/address-link/v/1",
```

Comment:
This can be `addressLinkTypeId`

Reply:
Thank you. 2af46146fb4f66ce7c7a30a794edd03721c029e5